diff --git a/containers/docker/status-alpine/geth/Dockerfile b/containers/docker/status-alpine/geth/Dockerfile
--- a/containers/docker/status-alpine/geth/Dockerfile	1970-01-01 01:00:00.000000000 +0100
+++ b/containers/docker/status-alpine/geth/Dockerfile	2017-09-19 17:03:51.000000000 +0200
@@ -0,0 +1,20 @@
+FROM alpine:3.5
+
+RUN \
+  apk add --update go git make gcc musl-dev linux-headers ca-certificates  && \
+
+  # clone status-go
+  mkdir -p /usr/lib/go/src/github.com/status-im && \
+  git clone --depth 1 --branch 0.9.7 https://github.com/status-im/status-go.git /usr/lib/go/src/github.com/status-im/status-go && \
+
+  # clone go-ethereum (and install everything)
+  git clone --depth 1 --branch status/1.6.1-stable https://github.com/status-im/go-ethereum && \
+  (cd go-ethereum && make geth) && \
+  cp go-ethereum/build/bin/geth /geth && \
+  apk del go git make gcc musl-dev linux-headers && \
+  rm -rf /go-ethereum && rm -rf /var/cache/apk/*
+
+EXPOSE 8545
+EXPOSE 30303
+
+ENTRYPOINT ["/geth"]
diff --git a/containers/docker/status-alpine/swarm/Dockerfile b/containers/docker/status-alpine/swarm/Dockerfile
--- a/containers/docker/status-alpine/swarm/Dockerfile	1970-01-01 01:00:00.000000000 +0100
+++ b/containers/docker/status-alpine/swarm/Dockerfile	2017-09-19 17:03:51.000000000 +0200
@@ -0,0 +1,19 @@
+FROM alpine:3.5
+
+RUN \
+  apk add --update go git make gcc musl-dev linux-headers ca-certificates  && \
+
+  # clone status-go
+  mkdir -p /usr/lib/go/src/github.com/status-im && \
+  git clone --depth 1 --branch develop https://github.com/status-im/status-go.git /usr/lib/go/src/github.com/status-im/status-go && \
+
+  # clone go-ethereum (and install everything)
+  git clone --depth 1 --branch status/1.6.1-stable https://github.com/status-im/go-ethereum && \
+  (cd go-ethereum && build/env.sh go run build/ci.go install ./cmd/swarm) && \
+  cp go-ethereum/build/bin/swarm /swarm && \
+  apk del go git make gcc musl-dev linux-headers && \
+  rm -rf /go-ethereum && rm -rf /var/cache/apk/*
+
+EXPOSE 30399
+
+ENTRYPOINT ["/swarm"]
diff --git a/containers/docker/status-alpine/wnode/Dockerfile b/containers/docker/status-alpine/wnode/Dockerfile
--- a/containers/docker/status-alpine/wnode/Dockerfile	1970-01-01 01:00:00.000000000 +0100
+++ b/containers/docker/status-alpine/wnode/Dockerfile	2017-09-19 17:03:51.000000000 +0200
@@ -0,0 +1,19 @@
+FROM alpine:3.5
+
+RUN \
+  apk add --update go git make gcc musl-dev linux-headers ca-certificates  && \
+
+  # clone status-go
+  mkdir -p /usr/lib/go/src/github.com/status-im && \
+  git clone --depth 1 --branch develop https://github.com/status-im/status-go.git /usr/lib/go/src/github.com/status-im/status-go && \
+
+  # clone go-ethereum (and install everything)
+  git clone --depth 1 --branch status/1.6.1-stable https://github.com/status-im/go-ethereum && \
+  (cd go-ethereum && build/env.sh go run build/ci.go install ./cmd/wnode) && \
+  cp go-ethereum/build/bin/wnode /wnode && \
+  apk del go git make gcc musl-dev linux-headers && \
+  rm -rf /go-ethereum && rm -rf /var/cache/apk/*
+
+EXPOSE 30379
+
+ENTRYPOINT ["/wnode"]
