diff --git a/eth/downloader/downloader.go b/eth/downloader/downloader.go
index eb64f05a..3e8c9cb1 100644
--- a/eth/downloader/downloader.go
+++ b/eth/downloader/downloader.go
@@ -230,6 +230,7 @@ func New(mode SyncMode, stateDb ethdb.Database, mux *event.TypeMux, chain BlockC
 		},
 		trackStateReq: make(chan *stateReq),
 	}
+	dl.cancelWg.Add(1)  // Add a sentinel to avoid getting flagged as a race condition in case Cancel is called before spawnSync
 	dl.downloads.Add(1) // Add a sentinel value to avoid waiting with zero count and getting a race condition signaled when terminating Downloader before starting sync
 	go dl.qosTuner()
 	go dl.stateFetcher()
 
